### ISSUE TITLE
How to train bounding box and polygon data together ?

### ISSUE THREAD

**Issue by Abhijeet241093** [ultralytics/ultralytics#19912]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello, thank you for the valuable resources. I have dataset contain bounding box and polygon mixed.

1.  How can I convert polygon dataset into bounding box  dataset? to trained  on YOLOv8 n/m series models. 
2.  How to train bounding box and polygon data together ? is it possible ? 

### Additional

Thank you

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Abhijeet241093, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Ultralytics Docs](https://docs.ultralytics.com/) for guidance on training, dataset preparation, and more. You'll find many [Python examples](https://docs.ultralytics.com/usage/python/) and [CLI examples](https://docs.ultralytics.com/usage/cli/) that might help address your questions.

To address your query:

1. If this is a ❓ Question, please share more details about your dataset, including examples of the bounding box and polygon annotations, and let us know the specific outcomes you're aiming for. This will help us provide more targeted advice.
2. If this is a 🐛 Bug Report, we kindly request a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further.

### Community Support
For additional help, consider joining the Ultralytics community:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time discussions
- [Discourse](https://community.ultralytics.com/) for detailed Q&A
- [Subreddit](https://reddit.com/r/Ultralytics) for knowledge sharing and discussions

### Upgrade
Please ensure you're using the latest version of the `ultralytics` package. Update it along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a Python>=3.8 environment with PyTorch>=1.8:

```bash
pip install -U ultralytics
```

### Verified Environments
YOLO can be run in several environments with all dependencies pre-installed:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon AWS**: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

### Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If the badge above is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests ensure YOLO runs smoothly across macOS, Windows, and Ubuntu for all [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/).

This is an automated response to help you get started. An Ultralytics engineer will review your issue and assist you further soon. Let us know if you need anything else in the meantime 😊!

**Comment by Abhijeet241093** [comment]
We annotate dataset using roboflow which contain both bounding box and polygon to represent wrinkle on dataset.

**Comment by glenn-jocher** [comment]
@Abhijeet241093 for converting polygons to bounding boxes, you can use the `segments2boxes()` function from the Ultralytics Simple Utilities. This converts segmentation polygons to `xywh` bounding box format directly. Here's a minimal example:

```python
from ultralytics.utils.ops import segments2boxes
segments = [...]  # your polygon data
boxes = segments2boxes(segments)
```

Regarding mixed training, YOLO models require task-specific formats (detection=boxes, segmentation=polygons). You'll need to either:
1. Convert all to detection format using the conversion tools in our Segmentation Datasets Guide
2. Convert boxes to segmentation format using `yolo_bbox2segment()` from the Simple Utilities

Choose one format based on your end task requirements. The models don't support simultaneous mixed-format training.

**Comment by ryouchinsa** [comment]
Hi @Abhijeet241093, roboflow is a great annotation tool. If your dataset contains bounding boxes and polygons, how about converting to the COCO format once. When you train bounding boxes, just convert the COCO format to the YOLO detection format, and when you train polygons, just convert the COCO format to the YOLO segmentation format.

```
"annotations": [
{
    "area": 164608,
    "bbox": [132, 417, 594, 432],
    "category_id": 14,
    "id": 1,
    "image_id": 1,
    "iscrowd": 0,
    "segmentation": [
        [136, 557, 152, 532, 191, 509, 266, 482, 367, 456, 375, 428, 427, 417, 486, 443, 516, 481, 518, 499, 564, 522, 611, 518, 661, 536, 701, 557, 724, 574, 719, 597, 691, 645, 723, 654, 715, 678, 695, 719, 681, 759, 681, 791, 670, 801, 659, 789, 656, 765, 627, 756, 644, 778, 629, 832, 591, 842, 553, 834, 514, 809, 494, 781, 491, 767, 433, 769, 403, 761, 405, 794, 387, 823, 369, 840, 344, 847, 309, 837, 295, 810, 286, 776, 290, 755, 297, 741, 259, 723, 216, 693, 179, 658, 147, 629, 132, 601, 132, 577]
    ]
},
```

**Comment by Abhijeet241093** [comment]
Yes, YOLOv8 OBB (Oriented Bounding Box) models support both oriented polygons and axis-aligned bounding boxes (AABB). That's better option.

**Comment by Abhijeet241093** [comment]
Just convert the COCO format to the YOLO detection format How to do this ?

**Comment by Abhijeet241093** [comment]
After getting bounding boxes in (x_min, y_min, x_max, y_max) format. should we have to convert into (cx, cy, w, h) format ?

**Comment by glenn-jocher** [comment]
@Abhijeet241093 for YOLO detection format, use the `convert_coco` function with `use_segments=False` to automatically convert COCO bbox annotations to normalized `xywh` format. The Ultralytics converter handles normalization and format conversion internally – no manual conversion needed. For implementation details, see the COCO to YOLO conversion guide in our documentation.

**Comment by Abhijeet241093** [comment]
I used this kind of logic, but not sure weather is correct or not ? Thank you

import os
from ultralytics.utils.ops import segments2boxes

label_dir = "train/labels" 
output_dir = "/label_converted"  

for file in os.listdir(label_dir):
    if file.endswith(".txt"):
        input_path = os.path.join(label_dir, file)
        output_path = os.path.join(output_dir, file)

        with open(input_path, "r") as f:
            lines = f.readlines()

        new_lines = []
        for line in lines:
            parts = line.strip().split()
            class_id = parts[0]  # Keep class_id as string
            values = [float(x) for x in parts[1:]]

            if len(values) == 4:
                # Already in YOLO bounding box format (x_center, y_center, width, height)
                new_lines.append(line)
            elif len(values) > 4:
                # Polygon format (convert to bounding box)
                segments = [values[i:i+2] for i in range(0, len(values), 2)]
                bbox = segments2boxes([segments])[0]  # Convert polygon to bbox (xywh)

                # Format output as class_id x_center y_center width height
                new_line = f"{class_id} {bbox[0]:.6f} {bbox[1]:.6f} {bbox[2]:.6f} {bbox[3]:.6f}\n"
                new_lines.append(new_line)

        with open(output_path, "w") as f:
            f.writelines(new_lines)

print("Conversion complete! Check:", output_dir)

**Comment by glenn-jocher** [comment]
Your approach is correct! The code properly converts polygon segments to YOLO bounding box format using `segments2boxes()`, preserves class IDs, and handles both existing boxes and polygons. For reference, see the [Segments2boxes Documentation](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.segments2boxes). Test with a sample file to verify output accuracy.

**Comment by glenn-jocher** [comment]
Great to hear it's working! 🚀 Feel free to share your accuracy results once ready, and we're here if you need further assistance. Keep up the great work with YOLO!

**Comment by Abhijeet241093** [comment]
Thank you very much @glenn-jocher

**Comment by glenn-jocher** [comment]
You're welcome! We're thrilled you're finding value in Ultralytics YOLO. The entire team and our amazing community appreciate your support. Let us know if you have any more questions as you continue your projects!

**Comment by Abhijeet241093** [comment]
![Image](https://github.com/user-attachments/assets/607f323a-83bc-45c1-b534-93403cd2b028)

![Image](https://github.com/user-attachments/assets/d97e042e-0b4f-494f-87d7-97dd680f0c0c)
![Image](https://github.com/user-attachments/assets/cd512ced-0618-448a-a91b-8ab8789b353b)
![Image](https://github.com/user-attachments/assets/4d567bb7-2e05-4f1d-94f0-7601b1c7a7eb)
![Image](https://github.com/user-attachments/assets/ac33fc02-a92e-410e-b423-585bf0830e83)

![Image](https://github.com/user-attachments/assets/ae924a7a-3d34-4cd3-96eb-a4f58804e6d8) 

Thank you for your guidance. @glenn-jocher. It's really helping.   !Is their any way to improve results for wrinkle?

**Comment by glenn-jocher** [comment]
Thanks for sharing your results, @Abhijeet241093! It's great to see the progress.

Improving detection results, especially for challenging classes like "wrinkle," often involves experimenting with a few key areas:

1.  **Data Augmentation:** Explore different augmentation techniques available during training. You can find details in the [Ultralytics Augmentation Guide](https://docs.ultralytics.com/integrations/albumentations/).
2.  **Hyperparameter Tuning:** Adjusting hyperparameters like learning rate, batch size, or the number of epochs can sometimes yield better performance. Our [Hyperparameter Tuning Guide](https://docs.ultralytics.com/guides/hyperparameter-tuning/) provides useful tips.
3.  **Model Size:** If you're using a smaller model (like YOLO11n), trying a larger variant (e.g., YOLO11s or YOLO11m) might capture finer details better, potentially improving wrinkle detection.
4.  **Dataset Quality:** Ensure your "wrinkle" annotations are consistent and accurately represent the features you want the model to learn. Class balance can also play a role; make sure you have sufficient examples of wrinkles.

Experimenting with these areas is often the best way to find improvements for specific classes. Good luck!

**Comment by Abhijeet241093** [comment]
It's amazing guidance.... Thank you very much, @glenn-jocher and the entire team!

**Comment by glenn-jocher** [comment]
You're very welcome, @Abhijeet241093! We're glad the guidance has been helpful. The Ultralytics team and the vibrant YOLO community are always here to support your efforts. Keep up the great work!
